nRF5340: DTS, SoC and Board definitions' re-work

ioannisg opened the pull request:
This PR extends the board definition for nRF5340, so both PDK and DK board targets are supported, along with some additional clean-up, without functional changes. In detail:
- in DTS
   -  we rename the nrf5340_cpuapp_common.dtsi to nrf5340_cpuapp_peripherals.dtsi
   - add some missing aliases
   - move the definition of the number of available MPU regions on the SoC header of the variant (instead of the common one)
   -  add the eng_a suffix in soc-variant-specific headers, to indicate that this variant concerns the Engineering A version

- in Board
   - rename the nRF5340 directory to nrf5340dk_nrf5340
   - remove the nrf5340pdk_ prefix from files that are to be shared between DK and PDK
   - add support for nrf5340dk

We also enable the ERRATUM19 on the PDK by default.

ioannisg committed:
dts: arm: nordic: rename nrf5340_cpuapp_common.dtsi

This commit renames the nrf5340_cpuapp_common.dtsi header
to nrf5340_cpuapp_peripherals.dtsi. This name is more
appropriate, as the header contains DT definitions for
peripherals, accessible from either Secure or Non-secure
execution environment (i.e. not accessible exclusively
by Secure or Non-Secure).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
dts: arm: nordic: move number of MPU regions in variant-specific file

Move the definition of the number of MPU regions
(num-mpu-regions) in a SoC variant-specific header
instead of the common .dtsi for nRF5340 Application
MCU. This allows us to define different SoC variants
with different number of MPU regions available.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
dts: arm: nordic: fix compatible strings for nRF5340 MCUs

We need to use CPUAPP and CPUNET instead of CPU0 and CPU1
terminology, because those terms are the ones used
throughout the DeviceTree definitions for nRF5340 SoC.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
dts: arm: nordic: add Eng A suffix to nRF5340 CPUAPP .dtsi headers

Rename the nRF5340 QKAA variant .dtsi headers (Application
MCU) by adding an Engineering A (Eng A) suffix. This is done
to indicate that the headers should be explicitly used when
building for a board that contains the Engineeering A nrf5340
SoC variant.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340pdk: remove un-needed nrf5340pdk prefix

We do not need to have the nrf5340pdk_ prefix in the
partition configuration and sram planning files for
nrf5340 platforms because these files may be common
for any nrf5340 board variant. Also, we do not need
it in the common .dts header for the Application MCU.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340: refactor common board Kconfig structure

Refactor the common Kconfig and build files of nRF5340
board definition to enable building for either nRF5340
PDK or DK.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340: rename nRF5340 board definition directory

We rename the nRF5340 board directory to
nrf5340dk_nrf5340 since this will be now
containing board definitions for both the
nRF5340 PDK and DK.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340: add definition for nRF5340 DK

Add the board definition for Nordic
nRF5340 DK.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340: enable erratum19 configuration on nRF5340 PDK

When building for nRF5340 PDK board, enable Kconfig option
for Erratum 19. Do not enable when building on nRF5340 DK.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
dts: arm: nordic: add SoC-variant headers for nrf5340 FP1

Adding SoC variant headers for nRF5340 variant (non Eng A).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340: remove SoC compatibles from board node

We do not need to list the SoC compatibles in the
PDK node definition, this was accidentally left out.
Removing it in this commit.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

kapi-no commented on boards/arm/nrf5340dk_nrf5340/Kconfig.board:
```diff
@@ -11,8 +11,18 @@ config BOARD_NRF5340PDK_NRF5340_CPUAPP
 config BOARD_NRF5340PDK_NRF5340_CPUAPPNS
 	bool "nRF5340 PDK nRF5340 Application MCU non-secure"
 
+config BOARD_NRF5340DK_NRF5340_CPUAPP
```
Could we provide Kconfig alias that does not distinguish between DK and PDK version of the board?

It could simplify a lot of if-endifs in this PR.

oyvindronningstad commented on boards/arm/nrf5340dk_nrf5340/nrf5340dk_nrf5340_cpunet.dts:
```diff
@@ -0,0 +1,145 @@
+&uart0 {
+	status = "okay";
+	current-speed = <115200>;
+	tx-pin = <25>;
+	rx-pin = <26>;
+	rts-pin = <10>;
+	cts-pin = <12>;
```
The back of my 0.9.0 board has

    P0.19 RTS
    P0.20 TXD
    P0.21 CTS
    P0.22 RXD

    P0.11 RTS
    P1.01 TXD
    P0.10 CTS
    P1.00 RXD

kapi-no replied:
If so, new UART pin settings need to be assigned to the Network Core by the Application Core in the cpunet_reset.c module. Unfortunately, we don't have access to the cpunet DTS in the cpuapp build targets.

kapi-no replied:
Please use UART pins for the Network Core that were listed by @oyvindronningstad. They are linked with the first COM port. This way, we will have two COM ports active - one for Application Core logs and one for Network Core logs. See my test commit for reference:

https://github.com/kapi-no/fw-nrfconnect-zephyr/commit/a0e6320accea67900d6fb273b7041b4d51e1239d

oyvindronningstad replied:
Note that I'm not sure what the pin info I listed means :P. I.e. which UARTs they refer to, and which core.

kapi-no replied:
Yes, but I think it is still sensible to leave the first set of UART pins for Application Core because pin configuration doesn't differ between the DK and PDK.

ioannisg replied:
>Please use UART pins for the Network Core that were listed by @oyvindronningstad. They are linked with the first COM port. This way, we will have two COM ports active - one for Application Core logs and one for Network Core logs.

@kapi-no done

kapi-no replied:
Thank you, please also update ``nrf5340_cpunet_reset.c `` file. Pin assignments still use the old UART configuration since they are not derived from DTS. Please see the linked commit in this discussion if you need further reference.

ioannisg replied:
Yes, I understand. Sure, I will have to change this too.

ioannisg replied:
Done now @kapi-no thanks for the input.

kapi-no replied:
Happy to help :) This PR looks good to me now, I have no further comments

anangl commented on dts/arm/nordic/nrf5340_cpuapp.dtsi:
```diff
@@ -85,7 +84,7 @@
 			 * peripheral description
 			 */
-			#include "nrf5340_cpuapp_common.dtsi"
+			#include "nrf5340_cpuapp_peripherals.dtsi"
```
Some of those peripherals (like WDT1, PWM3,  SPIM/SPIS/TWIM/TWIS/UARTE2-3) are not functional in the Eng A revision. I think we should delete their nodes in `*_eng_a.dtsi` files, to prevent them from being used.

ioannisg replied:
Agreed, @anangl , is this the ones I know about? I 'll add a separate commit and you can review it :)

carlescufi replied:
@anangl could you please provide @ioannisg with a list?

ioannisg replied:
> Could we change UART DTS pin configuration for Network Core within this PR?
> See:
> [#26110 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/26110#discussion_r444717827)
> for details

done, pls check if needed.

ioannisg replied:
applied now @anangl , on a separate commit on top.
Please note that I chose to "manually" delete the nodes in the ENG_A .dts file, instead of providing 2 file versions for nrf5340_cpuapp_peripherals, one for eng_a and one for fp1. I suppose this is an accepted way to do this.

mbolivar-nordic reviewed (approved):
I didn't check every line, but other than the specific issue reported by @anangl about aliases, DT changes seem reasonable to me.

emob-nordic commented on boards/arm/nrf5340dk_nrf5340/CMakeLists.txt:
```diff
@@ -1,7 +1,8 @@
 
-if ((CONFIG_BOARD_NRF5340PDK_NRF5340_CPUAPP OR CONFIG_BOARD_NRF5340PDK_NRF5340_CPUAPPNS)
+if (((CONFIG_BOARD_NRF5340PDK_NRF5340_CPUAPP OR CONFIG_BOARD_NRF5340PDK_NRF5340_CPUAPPNS)
+    OR (CONFIG_BOARD_NRF5340DK_NRF5340_CPUAPP OR CONFIG_BOARD_NRF5340DK_NRF5340_CPUAPPNS))
```
I am wondering if this bunch of CONFIG options could be replaced by `CONFIG_SOC_SERIES_NRF53X` ?
In fact each nrf5340(p)dk_nrf5340_cpuapp_defconfig enables `CONFIG_SOC_SERIES_NRF53X=y` so it could be simpler and cleaner. I think including this source is SoC specific not board specific. Could be similar for all over the place. What do You think?

carlescufi replied:
I agree, there has to be a better solution than this. If not the series, then maybe a new Kconfig option named `CONFIG_BOARD_NRF5340XDK` that is selected by both the PDK and DK boards.

ioannisg replied:
@carlescufi @emob-nordic what you are asking cannot be done, since either of the proposals, CONFIG_SOC_SERIES_NRF53X, or CONFIG_BOARD_NRF5340XDK would also be set when we build for the Network MCU.

The only way to clean this up would be to introduce APP/NET -specific, but DK-generic Kconfig options :) I was thinking this is not worth the effort, since we will probably not going to keep the PDK support in the Project, for too long. Eventually we will have only the "DK" support, and any common options we introduce will be very short-lived and could lead to more confusion. So I didn't want to do it, unless you insist.

emob-nordic replied:
I don't have strong opinion here but I thought it would be nice to shorten those `ifdefs` as they look almost the same and reader could easily misunderstand this.
If we plan to remove PDK support soon then Your argumentation seems logic for me and I would opt for it.
If we plan **not** to remove PDK support then I think introducing DK-genric Kconfig option is worthwhile.

PS: Do we really want to delete PDK support in the near future? I though its like "once in, never out".

ioannisg replied:
I think we will delete it as soon as we broadly introduce the final DKs

carlescufi commented on boards/arm/nrf5340dk_nrf5340/Kconfig:
```diff
@@ -43,9 +43,9 @@ config BOARD_ENABLE_CPUNET
+endif # BOARD_NRF5340PDK_NRF5340_CPUAPP || BOARD_NRF5340PDK_NRF5340_CPUAPPNS || BOARD_NRF5340DK_NRF5340_CPUAPP || BOARD_NRF5340DK_NRF5340_CPUAPPNS
 
-if BOARD_NRF5340PDK_NRF5340_CPUNET
+if BOARD_NRF5340PDK_NRF5340_CPUNET || BOARD_NRF5340DK_NRF5340_CPUNET
```
Same goes for those, can we make `BOARD_NRF5340XDK_NRF5340_CPUNET` and  `BOARD_NRF5340XDK_NRF5340_CPUAPP`?

ioannisg replied:
Right, this is an option, I just want us to evaluate, if this is really worth the effort, since we'll eventually clean up the PDK support.

carlescufi replied:
Yeah, I think we'll end up removing the support for the PDK as you say, I am convinced this is not worth it right now. Let's get this in as it is and if we want to clean it up later we can always do that.

carlescufi reviewed (changes requested):
In general I think the changes are fine, but we need to clean up the massive ifdefs.

carlescufi reviewed (approved):
Let's get this in as it is for now.

kapi-no commented:
Could we change UART DTS pin configuration for Network Core within this PR?
See:
https://github.com/zephyrproject-rtos/zephyr/pull/26110#discussion_r444717827
for details

ioannisg commented:
> Could we change UART DTS pin configuration for Network Core within this PR?
> See:
> [#26110 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/26110#discussion_r444717827)
> for details

done, pls, check if needed

ioannisg committed:
boards: nrf5340dk_nrf5340: properly choose UARTE pins for Network MCU

When building for nRF5340 Application MCU we need to properly
select which are the default GPIO port/pins for UARTE on the
Network MCU (so we allocate them to Network MCU). These  depend
on the actual version of the board we are using (PDK or DK) so
this commit implements the corresponding conditional logic.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
soc: arm: nrf5340: remove DTS nodes for non-functional devices

Several peripherals are non-functional in the ENG_A revision
of nRF5340 (Application MCU) so we delete the corresponding
DTS nodes in the device description.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
I'd merge this on @anangl 's approval.
